Add an example for  `InputChip` generated by user input

andreacioni opened the pull request:
New example for `InputChip` that demonstrate how to create/delete them based on user text inputs.

The sample application shows a custom text area where user can enter text. After the user has typed and hits _Enter_ the text will be replaced with an `InputChip` that contains that text. Is it possible to continue typing and add more chips in this way. All of them will be placed in a scrollable horizontal row. Also is it possible to have suggestion displayed below the text input field in case the typed text match some of the available suggestions.

Issue I'm trying to solve:

- https://github.com/flutter/flutter/issues/128247

**Code structure:**

The example app is composed of 2 main components that find places inside `MainScreen`:

 - `ChipsInput`
 - `ListView`

`ChipsInput` emulates a `TextField` where you can enter text. This text field accepts also a list of values of generic type T (`Topping` in my example), that gets rendered as `InputChip` inside the text field, before the text inserted by the user. This widgets is basically an `InputDecorator` widget that implements `TextInputClient` to get `TextEditingValue` events from the user keyboard. At the end of the input field there is another component, the `TextCursor`, that is displayed just when the user give the focus to the field and emulates the carrets that `TextField` has.

There are also some available callbacks that the user can use to capture events in the `ChipsInput` field like: `onChanged`, `onChipTapped`, `onSubmitted` and `onTextChanged`. This last callback is used to build a list of suggestion that will be placed just below the `ChipsInput` field inside the `ListView`.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

andreacioni committed:
new editable input chip field example

andreacioni committed:
add test

andreacioni committed:
Merge branch 'master' into master

TahaTesser commented:
Looks like you've a bunch of analyze issues 
https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8775518772474430497/+/u/Analyze/analyze/test_stdout

andreacioni commented:
Yeah 😅, I'm figuring these out. I'll try to fix them in a couple of days.

andreacioni committed:
fix linter errors

andreacioni committed:
add license

andreacioni committed:
removed print statements, changed replacement char

andreacioni committed:
Merge branch 'master' into master

andreacioni committed:
Merge branch 'master' into master

TahaTesser commented:
Hi @andreacioni 
Thanks for contributing a new example

1. Can you describe all the functionalities in this example in the [description](https://github.com/flutter/flutter/pull/130645#issue-1805732211)?
2. Describe the example code structure.
3. I can see visual misalignments on desktop
![Screenshot 2023-07-18 at 14 42 44](https://github.com/flutter/flutter/assets/48603081/06845883-3575-4dc0-8318-f8fdd5109383)
4. Backspace doesn't delete text on the desktop, just the web. Is this an example issue or Flutter itself?

andreacioni commented:
Hi @TahaTesser, Thanks for your time too. I'm really enjoying contributing to these project! 

I've updated the description but I've some questions related to the other points:

> Describe the example code structure.

Where it should be the best place to describe it? Here or the PR or in the code itself?

> I can see visual misalignments on desktop

Are you talking about the _Search_ hint text or the label?

> Backspace doesn't delete text on the desktop, just the web. Is this an example issue or Flutter itself?

Which OS are you using? Mac or Windows? I'm not able to reproduce on Mac OS X .

andreacioni committed:
Merge branch 'master' into master

TahaTesser commented:
> Where it should be the best place to describe it? Here or the PR or in the code itself?

In the [description](https://github.com/flutter/flutter/pull/130645#issue-1805732211), please explain the code structure. I can see there several things happening in `ChipsInput` and `MainScreen` 

> Are you talking about the Search hint text or the label?

Yes, please make sure there are no visual or functionality issues. 

 

> Which OS are you using? Mac or Windows? I'm not able to reproduce on Mac OS X .

Mac. Type some text and try to delete it with the backspace or delete key.

![Screenshot 2023-07-20 at 12 35 33](https://github.com/flutter/flutter/assets/48603081/ebd9847d-c3b2-4c72-9982-89198dbec3d1)

andreacioni commented:
Hi @TahaTesser,

some updates about:

> Mac. Type some text and try to delete it with the backspace or delete key.

I guess this is related to another [issue](https://github.com/flutter/flutter/issues/82124) that involves Mac OS X platform. I can't do much here if it didn't get resolved. The problem is related to the main component (`TextInputClient`) that I'm using in the example.

andreacioni committed:
fix hint alignment

give up using scrollable content because of the misalignment that caused on the hint

andreacioni committed:
Merge branch 'master' into master

andreacioni committed:
adjust spacing in Wrap widget

andreacioni commented:
Description is now updated. Let me know if you need more details.

TahaTesser commented on examples/api/lib/material/input_chip/input_chip.1.dart:
```diff
@@ -0,0 +1,484 @@
+  @override
+  TextCursorState createState() => TextCursorState();
+}
+
+class TextCursorState extends State<TextCaret>
+    with SingleTickerProviderStateMixin {
+  final Duration _duration = const Duration(milliseconds: 500);
+
+  bool _displayed = false;
+  late Timer _timer;
+
+  @override
+  void initState() {
+    super.initState();
+    _timer = Timer.periodic(_duration, _onTimer);
+  }
+
+  void _onTimer(Timer timer) {
+    setState(() => _displayed = !_displayed);
+  }
+
+  @override
+  void dispose() {
+    _timer.cancel();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final ThemeData theme = Theme.of(context);
+    return Opacity(
+      opacity: _displayed && widget.resumed ? 1.0 : 0.0,
+      child: Container(
+        width: 2.0,
+        height: 20,
+        color: theme.primaryColor,
+      ),
+    );
+  }
+}
```
```suggestion
// Copyright 2014 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'dart:async';

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

const List<String> _pizzaTopings = <String>[
  'Avocado',
  'Tomato',
  'Cheese',
  'Pepperoni',
  'Pickles',
  'Onion',
  'Jalapeno',
  'Mushrooms',
  'Pineapple',
];

void main() => runApp(const EditableChipFieldApp());

class EditableChipFieldApp extends StatelessWidget {
  const EditableChipFieldApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(useMaterial3: true),
      home: const EditableChipFieldExample(),
    );
  }
}

class EditableChipFieldExample extends StatefulWidget {
  const EditableChipFieldExample({super.key});

  @override
  EditableChipFieldExampleState createState() => EditableChipFieldExampleState();
}

class EditableChipFieldExampleState extends State<EditableChipFieldExample> {
  final FocusNode _chipFocusNode = FocusNode();
  Set<String> _toppings = <String>{ _pizzaTopings.first };
  List<String> _suggestions = <String>[];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Editable Chip Field Sample'),
      ),
      body: Column(
        children: <Widget>[
          Padding(
            padding: const EdgeInsets.symmetric(horizontal: 16),
            child: ChipsInput<String>(
              values: _toppings,
              decoration: const InputDecoration(
                prefixIcon: Icon(Icons.local_pizza_rounded),
                hintText: 'Search for toppings',
              ),
              focusNode: _chipFocusNode,
              onChanged: _onChanged,
              onSubmitted: _onSubmitted,
              chipBuilder: _chipBuilder,
              onTextChanged: _onSearchChanged,
            ),
          ),
          if (_suggestions.isNotEmpty)
            Expanded(
              child: ListView.builder(
                itemCount: _suggestions.length,
                itemBuilder: (BuildContext context, int index) {
                  return _suggestionBuilder(context, _suggestions[index]);
                },
              ),
            ),
        ],
      ),
    );
  }

  Future<void> _onSearchChanged(String value) async {
    final List<String> results = await _suggestionCallback(value);
    setState(() {
      _suggestions = results
        .where((String topping) => !_toppings.contains(topping))
        .toList();
    });
  }

  Widget _chipBuilder(BuildContext context, String topping) {
    return InputChip(
      key: ObjectKey(topping),
      label: Text(topping),
      avatar: CircleAvatar(
        child: Text(topping[0].toUpperCase()),
      ),
      onDeleted: () => _deleteChip(topping),
      onSelected: (_) => _onChipTapped(topping),
      materialTapTargetSize: MaterialTapTargetSize.shrinkWrap,
      padding: const EdgeInsets.all(2),
    );
  }

  Widget _suggestionBuilder(BuildContext context, String topping) {
    return ListTile(
      key: ObjectKey(topping),
      leading: CircleAvatar(
        child: Text(topping[0].toUpperCase()),
      ),
      title: Text(topping),
      onTap: () => _selectSuggestion(topping),
    );
  }

  void _selectSuggestion(String topping) {
    setState(() {
      _toppings.add(topping);
      _suggestions = <String>[];
    });
  }

  void _onChipTapped(String topping) {}

  void _deleteChip(String topping) {
    setState(() {
      _toppings.remove(topping);
      _suggestions = <String>[];
    });
  }

  void _onSubmitted(String text) {
    if (text.trim().isNotEmpty) {
      setState(() {
        _toppings = <String>{..._toppings, text.trim()};
      });
    } else {
      _chipFocusNode.unfocus();
      setState(() {
        _toppings = <String>{};
      });
    }
  }

  void _onChanged(Set<String> data) {
    setState(() {
      _toppings = data;
    });
  }

  FutureOr<List<String>> _suggestionCallback(String text) {
    if (text.isNotEmpty) {
      return _pizzaTopings.where((String topping) {
        return topping.toLowerCase().contains(text.toLowerCase());
      }).toList();
    }
    return const <String>[];
  }
}

class ChipsInput<T> extends StatefulWidget {
  const ChipsInput({
    super.key,
    required this.values,
    this.decoration = const InputDecoration(),
    required this.chipBuilder,
    required this.onChanged,
    this.onChipTapped,
    this.onSubmitted,
    this.onTextChanged,
    this.focusNode,
  });

  final Set<T> values;
  final InputDecoration decoration;
  final FocusNode? focusNode;

  final ValueChanged<Set<T>> onChanged;
  final ValueChanged<T>? onChipTapped;
  final ValueChanged<String>? onSubmitted;
  final ValueChanged<String>? onTextChanged;

  final Widget Function(BuildContext context, T data) chipBuilder;

  @override
  ChipsInputState<T> createState() => ChipsInputState<T>();
}

class ChipsInputState<T> extends State<ChipsInput<T>> implements TextInputClient {
  static const int kObjectReplacementChar = 0xFFFE;

  FocusNode? _lastChipFocusNode;

  late final FocusNode _focusNode;
  TextEditingValue _value = TextEditingValue.empty;
  TextInputConnection? _connection;

  @override
  void initState() {
    super.initState();
    _focusNode = widget.focusNode ?? FocusNode();
    _focusNode.addListener(_onFocusChanged);
  }

  @override
  void dispose() {
    _focusNode.dispose();
    _closeInputConnectionIfNeeded();
    super.dispose();
  }

  @override
  TextEditingValue get currentTextEditingValue => _value;

  @override
  void updateEditingValue(TextEditingValue value) {
    final int oldCount = _countReplacements(_value);
    final int newCount = _countReplacements(value);

    setState(() {
      if (newCount < oldCount) {
        widget.onChanged(widget.values.take(newCount).toSet());
      }
      _value = value;
    });
    widget.onTextChanged?.call(text);
  }

  @override
  void updateFloatingCursor(RawFloatingCursorPoint point) {}

  @override
  void connectionClosed() {
    _focusNode.unfocus();
  }

  @override
  void performAction(TextInputAction action) {
    widget.onSubmitted?.call(text);
  }

  @override
  AutofillScope? get currentAutofillScope => throw UnimplementedError();

  @override
  void didChangeInputControl(TextInputControl? oldControl, TextInputControl? newControl) {}

  @override
  void insertContent(KeyboardInsertedContent content) {}

  @override
  void insertTextPlaceholder(Size size) {}

  @override
  void performPrivateCommand(String action, Map<String, dynamic> data) {}

  @override
  void performSelector(String selectorName) {}

  @override
  void removeTextPlaceholder() {}

  @override
  void showAutocorrectionPromptRect(int start, int end) {}

  @override
  void showToolbar() {}

  @override
  Widget build(BuildContext context) {
    _syncTextValue();

    final List<Widget> chipsChildren = widget.values.map<Widget>(
      (T data) {
        final Widget child = widget.chipBuilder(context, data);
        if (_lastChipFocusNode != null) {
          _lastChipFocusNode!.requestFocus();
          return Focus(focusNode: _lastChipFocusNode, child: child);
        }
        return child;
      },
    ).toList();

    chipsChildren.add(Row(
      mainAxisSize: MainAxisSize.min,
      mainAxisAlignment: MainAxisAlignment.end,
      crossAxisAlignment: CrossAxisAlignment.end,
      children: <Widget>[
        Text(text),
        TextCaret(
          resumed: _focusNode.hasFocus,
        ),
      ],
    ));

    return GestureDetector(
      onTap: _requestKeyboard,
      child: InputDecorator(
        decoration: widget.decoration,
        isFocused: _focusNode.hasFocus,
        isEmpty: _value.text.isEmpty,
        textAlignVertical: TextAlignVertical.center,
        child: Wrap(
          crossAxisAlignment: WrapCrossAlignment.center,
          runSpacing: 4,
          spacing: 4,
          children: chipsChildren,
        ),
      ),
    );
  }

  String get text {
    return String.fromCharCodes(
      _value.text.codeUnits.where((int ch) => ch != kObjectReplacementChar),
    );
  }

  bool get _hasInputConnection => _connection?.attached ?? false;

  void _requestKeyboard() {
    if (_focusNode.hasFocus) {
      _openInputConnection();
    } else {
      FocusScope.of(context).requestFocus(_focusNode);
    }
  }

  void _onFocusChanged() {
    if (_focusNode.hasFocus) {
      _openInputConnection();
    } else {
      _closeInputConnectionIfNeeded();
    }
  }

  void _openInputConnection() {
    if (!_hasInputConnection) {
      _connection = TextInput.attach(this, const TextInputConfiguration());
      _connection!.setEditingState(_value);
    }
    _connection?.show();
  }

  void _closeInputConnectionIfNeeded() {
    if (_hasInputConnection) {
      _connection!.close();
      _connection = null;
    }
  }

  int _countReplacements(TextEditingValue value) {
    return _countReplacementsInString(value.text);
  }

  int _countReplacementsInString(String value) {
    return value.codeUnits
      .where((int ch) => ch == kObjectReplacementChar)
      .length;
  }

  int get currentReplacementCount => _countReplacements(_value);

  void _syncTextValue() {
    final TextEditingValue currentTextEditingValue = _value;
    final int newChipCount = widget.values.length;
    final int oldChipCount = _countReplacements(currentTextEditingValue);

    String text = String.fromCharCodes(List<int>.filled(newChipCount, kObjectReplacementChar));
    if (newChipCount == oldChipCount) {
      text += this.text;
    }
    _value = TextEditingValue(
      text: text,
      selection: TextSelection.collapsed(offset: text.length),
    );
    _connection?.setEditingState(_value);
  }
}

class TextCaret extends StatefulWidget {
  const TextCaret({super.key, this.resumed = false});

  final bool resumed;

  @override
  TextCursorState createState() => TextCursorState();
}

class TextCursorState extends State<TextCaret> with SingleTickerProviderStateMixin {
  final Duration _duration = const Duration(milliseconds: 500);
  bool _displayed = false;
  Timer? _timer;

  @override
  void initState() {
    super.initState();
    _timer = Timer.periodic(_duration, _onTimer);
  }

  void _onTimer(Timer timer) {
    setState(() {
      _displayed = !_displayed;
    });
  }

  @override
  void dispose() {
    _timer?.cancel();
    _timer = null;
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final ThemeData theme = Theme.of(context);
    return Opacity(
      opacity: _displayed && widget.resumed ? 1.0 : 0.0,
      child: Container(
        width: 2.0,
        height: 20,
        color: theme.primaryColor,
      ),
    );
  }
}
```

TahaTesser reviewed (changes requested):
I did some simplifications and cleanup.

TahaTesser commented:
This also requires some functionality tests and documentation.

See examples are documented in https://github.com/flutter/flutter/pull/128295/files

andreacioni committed:
Update examples/api/lib/material/input_chip/input_chip.1.dart

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

andreacioni committed:
typo and formatting fixes

andreacioni committed:
Merge branch 'master' into master

andreacioni commented:
Thanks for your feedback, really appreciated. I've included your changes along with some formatting and typo fixes. I'll take a look at the PR linked shortly.

andreacioni committed:
add example in the InputChip class documentations

andreacioni committed:
Merge branch 'master' into master

andreacioni commented:
@TahaTesser documentation updated! For feature test, do you suggest to expand `examples/api/test/material/input_chip/input_chip.1_test.dart` and add more test cases?

andreacioni committed:
Merge branch 'master' into master

TahaTesser commented:
> @TahaTesser documentation updated! For feature test, do you suggest to expand `examples/api/test/material/input_chip/input_chip.1_test.dart` and add more test cases?

The file you mentioned doesn't exist under:
https://github.com/flutter/flutter/tree/master/examples/api/test/material/input_chip

please create a matching test file and include tests.

andreacioni commented:
The file I was talking about is already part of the current PR: https://github.com/flutter/flutter/pull/130645/files#diff-2be583c1ca20bf6f76d1c1c7774d7b04517c58752eb39d9948f555fbf3261d32 . I wanted  to know if you want me to add more test cases inside it.

andreacioni committed:
Merge branch 'master' into master

TahaTesser commented:
> The file I was talking about is already part of the current PR: https://github.com/flutter/flutter/pull/130645/files#diff-2be583c1ca20bf6f76d1c1c7774d7b04517c58752eb39d9948f555fbf3261d32 . I wanted to know if you want me to add more test cases inside it.

Yes, a functionality test and a test that checks for the basic widget are found (if any main widgets are deleted, it will fail).  
The current test doesn't really check for anything but only the existence of the example.

andreacioni committed:
add more functional tests

andreacioni committed:
removed unused import

andreacioni committed:
Merge branch 'master' into master

andreacioni commented:
Test updated, basic functionality covered.

andreacioni committed:
Merge branch 'master' into master

TahaTesser commented on packages/flutter/lib/src/material/input_chip.dart:
```diff
@@ -42,6 +42,19 @@ import 'theme_data.dart';
 /// ** See code in examples/api/lib/material/input_chip/input_chip.0.dart **
 /// {@end-tool}
 ///
+/// Material design guidelines suggest that is possible to generate input chips
+/// from text input and vice versa. This is a common pattern in modern application
+/// and has many advantages like input verification and simplified field editing.
```
I am not sure this is needed.
```suggestion

```

TahaTesser commented on packages/flutter/lib/src/material/input_chip.dart:
```diff
@@ -42,6 +42,19 @@ import 'theme_data.dart';
+///
+///
+/// {@tool dartpad}
+/// The following example show you how to convert a text from
+/// user input to a list of one ore more [InputChip]. It cover also
+/// their removal and addition using a list of suggestions.
```
```suggestion
/// The following example showcases how to generate [InputChip]s from
/// user text input. When the user enters a pizza topping in the text field,
/// user is presented with a list of suggestions when selecting one of the
/// suggestions an [InputChip] is generated in the text field.
```

TahaTesser commented on examples/api/test/material/input_chip/input_chip.1_test.dart:
```diff
@@ -0,0 +1,58 @@
+  final String replacementChar =
+      String.fromCharCode(example.ChipsInputState.kObjectReplacementChar);
+
+  testWidgets('', (WidgetTester tester) async {
```
Add a test title

TahaTesser commented on examples/api/test/material/input_chip/input_chip.1_test.dart:
```diff
@@ -0,0 +1,58 @@
+    expect(find.byType(InputChip), findsNWidgets(2));
+
+    //simulate item deletion
+    await tester.tap(find.descendant(
+        of: find.byType(InputChip), matching: find.byType(InkWell).last));
```
```suggestion
    await tester.tap(find.descendant(
      of: find.byType(InputChip),
      matching: find.byType(InkWell).last,
    ));
```

TahaTesser commented on examples/api/test/material/input_chip/input_chip.1_test.dart:
```diff
@@ -0,0 +1,58 @@
+    await tester.pumpAndSettle();
+    expect(find.byType(InputChip), findsNWidgets(1));
+
+    await tester.tap(find.descendant(
+        of: find.byType(InputChip), matching: find.byType(InkWell).last));
```
```suggestion
    await tester.tap(find.descendant(
      of: find.byType(InputChip),
      matching: find.byType(InkWell).last,
    ));
```

TahaTesser commented on examples/api/test/material/input_chip/input_chip.1_test.dart:
```diff
@@ -0,0 +1,58 @@
+
+    example.ChipsInputState<String> state =
+        tester.state(find.byType(example.ChipsInput<String>));
+    expect(state.text, '');
```
Checking if `state.text` is empty?  Maybe try `isEmpty`?

andreacioni committed:
Update packages/flutter/lib/src/material/input_chip.dart

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

andreacioni committed:
Update examples/api/test/material/input_chip/input_chip.1_test.dart

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

andreacioni committed:
Update packages/flutter/lib/src/material/input_chip.dart

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

andreacioni committed:
Update examples/api/test/material/input_chip/input_chip.1_test.dart

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

andreacioni committed:
add test name and use isEmpty for text assertion

andreacioni committed:
Merge branch 'master' of https://github.com/andreacioni/flutter

andreacioni committed:
Merge branch 'master' into master

TahaTesser commented on packages/flutter/lib/src/material/input_chip.dart:
```diff
@@ -42,6 +42,17 @@ import 'theme_data.dart';
 /// ** See code in examples/api/lib/material/input_chip/input_chip.0.dart **
 /// {@end-tool}
 ///
+///
```
```suggestion
```

Remove extra line

TahaTesser commented on examples/api/test/material/input_chip/input_chip.1_test.dart:
```diff
@@ -0,0 +1,62 @@
+  final String replacementChar =
+      String.fromCharCode(example.ChipsInputState.kObjectReplacementChar);
+
+  testWidgets('Test EditableChipFieldApp', (WidgetTester tester) async {
```
```suggestion
  testWidgets('User input generates InputChips', (WidgetTester tester) async {
```

TahaTesser commented on examples/api/test/material/input_chip/input_chip.1_test.dart:
```diff
@@ -0,0 +1,62 @@
+    await tester.tap(find.byType(example.ChipsInput<String>));
+    await tester.pumpAndSettle();
+    expect(tester.testTextInput.isVisible, true);
+    // simulating text typing on input field
```
```suggestion
    // Simulating text typing on the input field.
```

TahaTesser commented on examples/api/test/material/input_chip/input_chip.1_test.dart:
```diff
@@ -0,0 +1,62 @@
+    await tester.pumpAndSettle();
+    expect(state.text, 'ham');
+
+    // add new InputChip by sending "done" action
```
```suggestion
    // Add new InputChip by sending the "done" action.
```

TahaTesser commented on examples/api/test/material/input_chip/input_chip.1_test.dart:
```diff
@@ -0,0 +1,62 @@
+
+    expect(find.byType(InputChip), findsNWidgets(2));
+
+    //simulate item deletion
```
```suggestion
    // Simulate item deletion.
```

TahaTesser commented on examples/api/test/material/input_chip/input_chip.1_test.dart:
```diff
@@ -0,0 +1,62 @@
+      matching: find.byType(InkWell).last,
+    ));
+    await tester.pumpAndSettle();
+    expect(find.byType(InputChip), findsNWidgets(0));
```
```suggestion
    expect(find.byType(InputChip), findsNothing);
```

TahaTesser commented on examples/api/test/material/input_chip/input_chip.1_test.dart:
```diff
@@ -0,0 +1,62 @@
+    );
+    await tester.pumpAndSettle();
+
+    expect(find.byType(example.EditableChipFieldApp), findsNWidgets(1));
```
```suggestion
    expect(find.byType(example.EditableChipFieldApp), findsOneWidget);
```

TahaTesser replied:
If we're finding one widget, we don't need to NWidgets

TahaTesser commented on examples/api/test/material/input_chip/input_chip.1_test.dart:
```diff
@@ -0,0 +1,62 @@
+
+    expect(find.byType(example.EditableChipFieldApp), findsNWidgets(1));
+    expect(find.byType(example.ChipsInput<String>), findsNWidgets(1));
+    expect(find.byType(InputChip), findsNWidgets(1));
```
```suggestion
    expect(find.byType(InputChip), findsOneWidget);
```

TahaTesser commented on examples/api/test/material/input_chip/input_chip.1_test.dart:
```diff
@@ -0,0 +1,62 @@
+    await tester.pumpAndSettle();
+
+    expect(find.byType(example.EditableChipFieldApp), findsNWidgets(1));
+    expect(find.byType(example.ChipsInput<String>), findsNWidgets(1));
```
```suggestion
    expect(find.byType(example.ChipsInput<String>), findsOneWidget);
```

TahaTesser commented on examples/api/test/material/input_chip/input_chip.1_test.dart:
```diff
@@ -0,0 +1,62 @@
+    // simulating text typing on input field
+    tester.testTextInput.enterText('${replacementChar}ham');
+    await tester.pumpAndSettle();
+    expect(find.byType(InputChip), findsNWidgets(1));
```
```suggestion
    expect(find.byType(InputChip), findsOneWidget);
```

TahaTesser commented on examples/api/test/material/input_chip/input_chip.1_test.dart:
```diff
@@ -0,0 +1,62 @@
+    await tester.pumpAndSettle();
+    expect(state.text.isEmpty, true);
+
+    expect(find.byType(InputChip), findsNWidgets(2));
```
This is correct since we're finding more than one widget.

TahaTesser commented on examples/api/test/material/input_chip/input_chip.1_test.dart:
```diff
@@ -0,0 +1,62 @@
+      matching: find.byType(InkWell).last,
+    ));
+    await tester.pumpAndSettle();
+    expect(find.byType(InputChip), findsNWidgets(1));
```
```suggestion
    expect(find.byType(InputChip), findsOneWidget);
```

TahaTesser commented:
@andreacioni 
Last small changes. This looks good.

andreacioni committed:
Update packages/flutter/lib/src/material/input_chip.dart

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

andreacioni committed:
Update examples/api/test/material/input_chip/input_chip.1_test.dart

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

andreacioni committed:
Update examples/api/test/material/input_chip/input_chip.1_test.dart

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

andreacioni committed:
Update examples/api/test/material/input_chip/input_chip.1_test.dart

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

andreacioni committed:
Update examples/api/test/material/input_chip/input_chip.1_test.dart

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

andreacioni committed:
Update examples/api/test/material/input_chip/input_chip.1_test.dart

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

andreacioni committed:
Update examples/api/test/material/input_chip/input_chip.1_test.dart

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

andreacioni committed:
Update examples/api/test/material/input_chip/input_chip.1_test.dart

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

andreacioni committed:
Apply suggestions from code review

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

andreacioni committed:
Merge branch 'master' into master

andreacioni committed:
Merge branch 'master' into master

TahaTesser reviewed (approved):
LGTM! 

Ths will need a second review before we land.

HansMuller commented on examples/api/lib/material/input_chip/input_chip.1.dart:
```diff
@@ -0,0 +1,434 @@
+  }
+}
+
+class TextCaret extends StatefulWidget {
```
This seems unnecessarily unconventional.

The state class is (conventionally):
```dart
class _TextCaretState extends State<TextCaret>
```
and so the TextCaret override would be:
```dart
State<TextCaret> createState() => _TextCaretState();
```

HansMuller commented on examples/api/lib/material/input_chip/input_chip.1.dart:
```diff
@@ -0,0 +1,434 @@
+              child: ListView.builder(
+                itemCount: _suggestions.length,
+                itemBuilder: (BuildContext context, int index) {
+                  return _suggestionBuilder(context, _suggestions[index]);
```
Prefer to factor widgets with more widgets, so:
```dart
return ToppingSuggestion(_suggestions[index]);
```

HansMuller commented on examples/api/lib/material/input_chip/input_chip.1.dart:
```diff
@@ -0,0 +1,434 @@
+}
+
+class ChipsInputState<T> extends State<ChipsInput<T>>
+    implements TextInputClient {
```
@justinmc - is there a simpler way to do this than implementing TextInputClient?

justinmc replied:
Yeah we shouldn't be recommending people do this in our examples, because this will result in the loss of lots of functionality in EditableText.  For example, the delete key doesn't work in this example with a hardware keyboard.

I think the best way to do this would be to extend TextEditingController and override buildTextSpan and then use a TextField like normal.  You can insert your chips inline into the text as WidgetSpans in buildTextSpan.

CC @Renzo-Olivares who is the real expert in this.

justinmc reviewed (commented):
We'll have to get rid of the `implements TextInputClient` in this example as discussed in a comment below.  But this is a pretty important thing to provide an example for, so thank you for taking it on.  I'm happy to help advise on the TextInputClient thing if it gets tricky.

TahaTesser commented on examples/api/lib/material/input_chip/input_chip.1.dart:
```diff
@@ -0,0 +1,434 @@
+  'Tomato',
+  'Cheese',
+  'Pepperoni',
+  'Pickles',
```
I might have suggested that haha. I'm no pizza expert.

justinmc replied:
I guess I should hold off on trying it for now 😄

andreacioni replied:
Yeah definitely not the best toppings I'd suggest 😅, let's leave it there for now for sake of simplicity.

andreacioni commented:
@justinmc thanks for your hint on TextEditingController and buildTextSpan. It sure worth a try because it could simplify a lot. I should have some spare time to work on it before merging.

andreacioni committed:
replaced TextInputClient

andreacioni committed:
Merge branch 'master' into master

andreacioni committed:
Merge branch 'master' into master

andreacioni committed:
Merge branch 'master' into master

andreacioni committed:
removed _suggestionBuilder method

andreacioni committed:
add textInputAction

andreacioni committed:
Merge branch 'master' into master

andreacioni committed:
Merge branch 'master' into master

andreacioni committed:
refactor buildSpan

andreacioni committed:
Merge branch 'master' of https://github.com/andreacioni/flutter

andreacioni commented:
@HansMuller, @justinmc It took longer than expected but now the new `InputChip` example is completed. I’ve used both `TextField` and the extended `TextEditingController` as you suggested. I can say that you were right: much more easier to build and understand.

**Just a side note**, not sure if it could interest, but my previous test failed on flutter channel master and passed on stable because of this line:

```
return TextSpan(style: style, children: <InlineSpan>[
      ...chipWidgets,
      if (textWithoutReplacements.isNotEmpty)// without the "if" the test fails on master and pass on stable
        TextSpan(text: textWithoutReplacements)
    ]);
```

the error appears when you try to the tap on delete icon when there is just one `InputChip`.

andreacioni committed:
input chip builder as separate widget

andreacioni committed:
Merge branch 'master' into master

justinmc reviewed (approved):
LGTM with nits :+1: 

This looks like a very good example to give to users looking to build a custom input like this.  CC @Renzo-Olivares in case you have any thoughts about the pattern of extending a TextEditingController like this.

andreacioni committed:
minor fixes

andreacioni committed:
Merge branch 'master' of https://github.com/andreacioni/flutter

andreacioni committed:
Update packages/flutter/lib/src/material/input_chip.dart

Co-authored-by: Justin McCandless <justinjmccandless@gmail.com>

andreacioni committed:
Update examples/api/lib/material/input_chip/input_chip.1.dart

Co-authored-by: Justin McCandless <justinjmccandless@gmail.com>

andreacioni committed:
Merge branch 'master' into master

andreacioni committed:
Merge branch 'master' into master

andreacioni committed:
Merge branch 'master' into master

justinmc reviewed (approved):
Renewing my LGTM, thanks for fixing my nits.

HansMuller commented on examples/api/lib/material/input_chip/input_chip.1.dart:
```diff
@@ -0,0 +1,367 @@
+      style: widget.style,
+      strutStyle: widget.strutStyle,
+      controller: controller,
+      onChanged: (_) =>
```
Please include parameters and types, even if they're not used per https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-anonymous-parameter-names

HansMuller commented on examples/api/lib/material/input_chip/input_chip.1.dart:
```diff
@@ -0,0 +1,367 @@
+  const EditableChipFieldExample({super.key});
+
+  @override
+  EditableChipFieldExampleState createState() =>
```
When defining a function with `=>`, put the entire definition on one line. Here and elsewhere.

HansMuller commented on examples/api/lib/material/input_chip/input_chip.1.dart:
```diff
@@ -0,0 +1,367 @@
+      {required BuildContext context,
+      TextStyle? style,
+      required bool withComposing}) {
+    final Iterable<WidgetSpan> chipWidgets =
```
One line

andreacioni replied:
I'm not sure how to do that, the auto formatter is preventing me to do so. Should I disable it to achieve the one-line syntax? Sorry for the silly question I've never had to do that.

HansMuller commented on examples/api/lib/material/input_chip/input_chip.1.dart:
```diff
@@ -0,0 +1,367 @@
+}
+
+class ToppingInputChip extends StatelessWidget {
+  const ToppingInputChip(
```
Formatting:
```dart
  const ToppingInputChip({
    super.key,
    required this.topping,
    required this.onDeleted,
    required this.onSelected
  });
```

HansMuller reviewed (commented):
This is great, just some random formatting things to smooth out.

andreacioni commented on examples/api/lib/material/input_chip/input_chip.1.dart:
```diff
@@ -0,0 +1,367 @@
+
+  static int countReplacements(String text) {
+    return text.codeUnits
+        .where(
```
The standard formatter prevents me to place it on a single line. I'm going to remove the arrow function and replaced with a standard anonymous one.

andreacioni committed:
fixing format

andreacioni committed:
missing value type in a callback

andreacioni committed:
Merge branch 'master' into master

andreacioni committed:
Merge branch 'master' into master

justinmc commented:
@andreacioni I think you need to disable your autoformatter, flutter/flutter does not use dartfmt: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#formatting

andreacioni committed:
foirced one line format

andreacioni committed:
Merge branch 'master' into master

andreacioni committed:
removed trailing whitespaces

andreacioni committed:
Merge branch 'master' of https://github.com/andreacioni/flutter

andreacioni committed:
Merge branch 'master' into master

HansMuller commented on packages/flutter/lib/src/material/input_chip.dart:
```diff
@@ -42,6 +42,16 @@ import 'theme_data.dart';
 ///
+///
+/// {@tool dartpad}
+/// The following example showcases how to generate [InputChip]s from
```
```suggestion
/// The following example shows how to generate [InputChip]s from
```

HansMuller reviewed (approved):
LGTM

HansMuller committed:
Update packages/flutter/lib/src/material/input_chip.dart

HansMuller committed:
Merge branch 'master' into master

HansMuller committed:
Merge branch 'master' into master

HansMuller commented:
NICE

![Screenshot 2023-09-01 at 12 17 58 PM](https://github.com/flutter/flutter/assets/1377460/8767991b-7140-49cd-9640-63094ea1cedb)
